index/index.go

```diff
@@ -193,59 +121,98 @@ func (index *Index) addFile(subPath string) error {
 	return nil
 }
 
-func newIndexer(path string, fileIndex int, index *Index) indexer {
-	indexer := indexer{
-		index:     index,
-		fileIndex: fileIndex,
-		path:      path,
+func (index *Index) buildIndexOverFile(fileId int, f indexFile) error {
+	entry := make([]byte, 12)
+
+	for {
+		offset, err := f.file.Seek(0, 1)
+		if err != nil {
+			return err
+		}
+
+		if !f.reader.ScanKey() {
+			break
+		}
+
+		serializeIndexEntry(entry, fileId, offset)
+		index.ldb.Put(f.reader.Key(), entry, nil)
+		index.count++
+	}
+
+	if err := f.reader.Err(); err != nil {
+		return err
 	}
 
-	return indexer
+	log.Println("Finished indexing", f.file.Name())
+	return nil
 }
 
-func (i *indexer) start() {
-	file, err := os.Open(i.path)
+// Get returns the value for a given key.
+func (index *Index) Get(key string) ([]byte, error) {
+	if !index.Ready {
+		return nil, errors.New("Index isn't finished being built yet.")
+	}
+
+	bytes, err := index.ldb.Get([]byte(key), nil)
+	if err == leveldb.ErrNotFound {
+		return nil, ErrNotFound
+	} else if err != nil {
+		return nil, err
+	}
+
+	if len(bytes) != 12 {
+		return nil, fmt.Errorf("Invalid value length: %d", len(bytes))
+	}
+
+	fileId, offset := deserializeIndexEntry(bytes)
+	f := index.files[fileId]
+
+	readLock := index.readLocks[fileId]
```

Comment:
(this is also existing code)

Yeah, as far as I can tell, the read path isn't really the problem. But it might be an interesting optimization to use a `sync.Pool` for each file, or something.